src/frontend/src/flows/register/captcha.ts

```diff
@@ -39,108 +39,101 @@ export const promptCaptchaTemplate = <T>({
 }) => {
   const copy = i18n.i18n(copyJson);
 
+  const spinnerImg: TemplateResult = html`
+    <div class="c-captcha-placeholder c-spinner-wrapper">
+      <div class="c-spinner">${spinner}</div>
+    </div>
+  `;
+  const captchaImg = (base64: string): TemplateResult =>
+    html`<div class="c-captcha-placeholder">
+      <img
+        src="data:image/png;base64,${base64}"
+        id="captchaImg"
+        class="c-image"
+        alt="captcha image"
+      />
+    </div>`;
+
+  // The various states the component can inhabit
+  type State =
+    | { status: "requesting" }
+    | { status: "prompting"; challenge: Challenge }
+    | { status: "verifying" }
+    | { status: "bad" };
+
   // We define a few Chans that are used to update the page in a
-  // reactive way; see template returned by this function
+  // reactive way based on state; see template returned by this function
+  const state = new Chan<State>({ status: "requesting" });
 
   // The image shown
-  const img = new Chan<TemplateResult>();
+  const img: Chan<TemplateResult> = state.map({
+    f: (state) =>
+      state.status === "requesting"
+        ? spinnerImg
+        : state.status === "prompting"
+        ? captchaImg(state.challenge.png_base64)
+        : Chan.unchanged,
+    def: spinnerImg,
+  });
 
   // The text input where the chars can be typed
   const input: Ref<HTMLInputElement> = createRef();
 
   // The error shown on bad input
-  const errorText = new Chan<DynamicKey | undefined>();
-  const hasError = errorText.map((e) => (e !== undefined ? "has-error" : ""));
+  const errorText = state.map(({ status }) =>
+    status === "bad" ? copy.incorrect : undefined
+  );
+  const hasError = state.map(({ status }) =>
+    status === "bad" ? "has-error" : ""
+  );
 
   // The "next" button behavior
-  const next = new Chan<((e: SubmitEvent) => void) | undefined>();
-  const nextDisabled = next.map((f) => f === undefined);
-  const nextCaption = new Chan<DynamicKey>();
+  const next: Chan<((e: SubmitEvent) => void) | undefined> = state.map(
+    (state) =>
+      state.status === "prompting"
+        ? (e) => {
+            e.preventDefault();
+            e.stopPropagation();
+            doVerify(state.challenge);
+          }
+        : undefined
+  );
 
-  // The "retry" button behavior
-  const retry = new Chan<(() => void) | undefined>();
-  const retryDisabled = retry.map((f) => f === undefined);
+  const nextDisabled: Chan<boolean> = next.map((f) => f === undefined);
+  const nextCaption: Chan<DynamicKey> = state.map(({ status }) =>
+    status === "requesting"
+      ? copy.generating
```

Comment:
it's correct that "requesting" map to "generating"? just in case not expected

Reply:
yep, we're requesting a new captcha but the copy says "generating". Could make them both the same but don't want to have to update the state name every time the copy changes. Thanks for paying attention though!